controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,289 @@
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	pod, err := kc.pLister.Pods(namespace).Get(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Pod: %s", name)
+	}
+	nodeID := pod.Spec.NodeName
+	if err := kc.handlePodDeletionIfNodeDown(pod, nodeID, namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// handlePodDeletionIfNodeDown determines whether we are allowed to forcefully delete a pod
+// from a failed node based on the users chosen NodeDownPodDeletionPolicy.
+// This is necessary because Kubernetes never forcefully deletes pods on a down node,
+// the pods are stuck in terminating state forever and Longhorn volumes are not released.
+// We provide an option for users to help them automatically force delete terminating pods
+// of StatefulSet/Deployment on the downed node. By force deleting, k8s will detach Longhorn volumes
+// and spin up replacement pods on a new node.
+//
+// Force delete a pod when all of the below conditions are meet:
+// 1. NodeDownPodDeletionPolicy is different than DoNothing
+// 2. pod belongs to a StatefulSet/Deployment depend on NodeDownPodDeletionPolicy
+// 3. node containing the pod is down
+// 4. the pod is terminating and the DeletionTimestamp has passed.
+// 5. pod has a PV with provisioner driver.longhorn.io
+func (kc *KubernetesPodController) handlePodDeletionIfNodeDown(pod *v1.Pod, nodeID string, namespace string) error {
+	deletionPolicy := types.NodeDownPodDeletionPolicyDoNothing
+	if deletionSetting, err := kc.ds.GetSettingValueExisted(types.SettingNameNodeDownPodDeletionPolicy); err == nil {
+		deletionPolicy = types.NodeDownPodDeletionPolicy(deletionSetting)
+	}
+
+	shouldDelete := (deletionPolicy == types.NodeDownPodDeletionPolicyDeleteStatefulSetPod && isOwnedByStatefulSet(pod)) ||
+		(deletionPolicy == types.NodeDownPodDeletionPolicyDeleteDeploymentPod && isOwnedByDeployment(pod)) ||
+		(deletionPolicy == types.NodeDownPodDeletionPolicyDeleteBothStatefulsetAndDeploymentPod && (isOwnedByStatefulSet(pod) || isOwnedByDeployment(pod)))
+
+	if !shouldDelete {
+		return nil
+	}
+
+	isNodeDown, err := kc.ds.IsNodeDownOrDeleted(nodeID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to evaluate Node %v for pod %v in handlePodDeletionIfNodeDown", nodeID, pod.Name)
+	}
+	if !isNodeDown {
+		return nil
+	}
+
+	if pod.DeletionTimestamp == nil || pod.DeletionTimestamp.After(time.Now()) {
+		return nil
+	}
+
+	gracePeriod := int64(0)
+	err = kc.kubeClient.CoreV1().Pods(namespace).Delete(pod.Name, &metav1.DeleteOptions{
+		GracePeriodSeconds: &gracePeriod,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to forcefully delete Pod %v on the downed Node %v in handlePodDeletionIfNodeDown", pod.Name, nodeID)
+	}
+	logrus.Infof("%v: Forcefully deleted pod %v on downed node %v", controllerAgentName, pod.Name, nodeID)
```

Comment:
Would be good to keep messages consistent, suggest dropping the `controllerAgentName` here too.

Reply:
same as https://github.com/longhorn/longhorn-manager/pull/640#discussion_r476965100